lib/rmt/downloader.rb

```diff
@@ -89,72 +83,99 @@ def create_fiber_request(local_filenames, file_reference, failed_downloads: nil)
         @logger.warn("× #{File.basename(file_reference.local_path)} - #{e}")
         failed_downloads << file_reference.local_path
       ensure
-        process_queue(local_filenames, failed_downloads)
+        process_queue(failed_downloads)
       end
     end
 
     request_fiber.resume
   end
 
-  def process_queue(local_filenames, failed_downloads = nil)
+  def process_queue(failed_downloads = nil)
     queue_item = @queue.shift
     return unless queue_item
 
-    @hydra.queue(
-      create_fiber_request(
-        local_filenames,
-        queue_item,
-        failed_downloads: failed_downloads
-      )
-    )
+    request = create_fiber_request(queue_item, failed_downloads: failed_downloads)
+    @hydra.queue(request)
   end
 
   def make_request(file, request_fiber)
-    uri = URI.join(file.remote_path)
-    uri.query = @auth_token if (@auth_token && uri.scheme != 'file')
-
-    if URI(uri).scheme == 'file' && !File.exist?(uri.path)
-      raise RMT::Downloader::Exception.new(_('%{file} - File does not exist') % { file: file.remote_path })
-    end
-
     downloaded_file = Tempfile.new('rmt', Dir.tmpdir, mode: File::BINARY, encoding: 'ascii-8bit')
 
-    headers = {}
-    headers['If-Modified-Since'] = file.cache_timestamp if file.cache_timestamp
-
     request = RMT::FiberRequest.new(
-      uri.to_s,
+      request_uri(file).to_s,
       download_path: downloaded_file,
       request_fiber: request_fiber,
-      remote_file: file.remote_path,
-      followlocation: true,
-      headers: headers
+      followlocation: true
     )
 
     request.receive_headers
     request.receive_body
   end
 
+  def try_copying_from_cache(files, ignore_errors: false)
+    cacheable_files = files.group_by { |file| !file.cache_timestamp.nil? }
+    uncached_files = cacheable_files.fetch(false, [])
+    cacheable_files = cacheable_files.fetch(true, [])
+      .map { |file| [file, head_request(file)] }.to_h
+
+    return [uncached_files, []] if cacheable_files.empty?
+
+    if cacheable_files.count > 1
+      hydra = Typhoeus::Hydra.new(max_concurrency: @concurrency)
+      cacheable_files.each_value { |request| hydra.queue(request) }
+      hydra.run
+    else
+      cacheable_files.values.first.run
+    end
+
+    failed_files = []
+    cacheable_files.each do |(file, request)|
+      next uncached_files << file unless valid_cached_file?(file, request.response)
+
+      copy_from_cache(file)
+    rescue RMT::Downloader::Exception => e
+      next failed_files << file.local_path if ignore_errors
+
+      raise e
+    end
+
+    [uncached_files, failed_files]
+  end
+
+  def head_request(file)
+    RMT::HttpRequest.new(request_uri(file).to_s, method: :head, followlocation: true)
```

Comment:
Do you think that we need to guard this function to ensure the head request is for http/https?